File: code-search/python/lcbbc.py
#!/usr/bin/env python

import sys
import math

DEBUG = True
def log(msg):
	if DEBUG:
		print msg

def usage():
	sys.stderr.write("%s: <n> <min_hd>\n" % (sys.argv[0]))


MAX_N = 8
MAX_CAND = (1 << MAX_N)

HAMMING_WEIGHT = [0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4, 1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4, 4, 5, 1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4, 4, 5, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4, 4, 5, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4, 4, 5, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 2, 3, 3, 4, 3, 4, 4, 5, 3, 4, 4, 5, 4, 5, 5, 6, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 4, 5, 5, 6, 5, 6, 6, 7, 5, 6, 6, 7, 6, 7, 7, 8]
HD = []

# fill big lookup table of hamming distances
def precompute_hd(min_hd):
	for i in range(MAX_CAND):
		HD.append([])
		for j in range(MAX_CAND):
			if HAMMING_WEIGHT[i ^ j] >= min_hd:
				HD[i].append(True)
			else:
				HD[i].append(False)


'''
Algorithm
  find list of candidates for codes assuming start from 0 for each candidate in
  the candidate set, check hamming dist to remaining candidates eventually
  producing a code known to obey min_hd.
  
  backtrack to find other possible codes with greater length
'''

def populate_candidates(code, candidates):
	return filter(lambda x: HD[code][x], candidates)

def iterative_find_code(n):
	# One time setup
	code = [0]
	longest_code = 1
	search_space = range((1<<n)-1, 0, -1)
	candidates = populate_candidates(code[0], search_space)
	cand_stack = [candidates]
	results = []
	
	while cand_stack:
		candidates = cand_stack[-1]
		if not len(candidates):
			cand_stack.pop()
			code.pop()
			continue

		codeword = candidates.pop()
		code.append(codeword)
		next_candidates = populate_candidates(codeword, candidates)
	
		if next_candidates and len(next_candidates) + len(code) > longest_code:
			cand_stack.append(next_candidates)
		else:
			if len(code) > longest_code:
				results.append(code[:])
				log("%d, %f, " % (len(code), math.log(len(code), 2)))
				log(code)
				longest_code = len(code)
			code.pop()
	return results


if __name__ == '__main__':
	if len(sys.argv) != 3:
		usage()
		sys.exit(1)

	n, min_hd = map(int, sys.argv[1:])
	precompute_hd(min_hd)

	results = iterative_find_code(n)
	#for code in results:
	#	log(code)

File: code-search/c/lcbbc.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include "TRACE.h"

#define MAX_N    8
#define MAX_CAND (1 << MAX_N)

uint8_t HD[MAX_CAND][MAX_CAND];

void usage(char *argv0) {
	fprintf(stderr, "%s: <n> <min_hd>\n", argv0);
	fprintf(stderr, "    MAX_N = %d\n", MAX_N);
}

/* fill big lookup table of hamming distances */
void precompute_hd(int min_hd)
{
	int i, j;
	for (i = 0; i < MAX_CAND; i++)
		for (j = 0; j < MAX_CAND; j++)
			/* FIXME store as bits in bytes and
			 * use popcount to remove MAX_N limitation
			 */
			if(__builtin_popcount(i ^ j) >= min_hd)
				HD[i][j] = 1;
			else
				HD[i][j] = 0;
}

typedef struct codeword_list_t {
	uint16_t index;
	uint16_t codewords[MAX_CAND];
} codeword_list;

void populate_candidates(uint16_t code, codeword_list *candidates,
						 codeword_list* next_candidates)
{
	uint16_t nc;
	int i;
	for (i = 0; i < candidates->index; i++) {
		nc = candidates->codewords[i];
		if (HD[code][nc])
			next_candidates->codewords[next_candidates->index++] = nc;
	}
}

/* recursive search for longest codes */
int find_best_code(codeword_list* code, codeword_list* candidates, int longest)
{
	codeword_list next_candidates;
	uint16_t c;

	while (candidates->index) {
		candidates->index--;
		c = candidates->codewords[candidates->index];
		code->codewords[code->index++] = c;
		next_candidates.index = 0;
		populate_candidates(c, candidates, &next_candidates);

		if ((next_candidates.index + code->index) > longest) {
			if(next_candidates.index)
				longest = find_best_code(code, &next_candidates, longest);
			else {
				longest = code->index;
				printf("%d, [", longest);
				int i;
				for(i=0; i<longest; i++)
					printf("%d, ", code->codewords[i]);
				printf("]\n");
			}
		}
		code->index--;
	}
	return longest;
}

int find_code(int bitlength) {
	int i;
	codeword_list code, candidates;
	code.index = 1;
	code.codewords[0] = 0;
	candidates.index = 1 << bitlength;
	for (i = 1; i <= candidates.index; i++)
		candidates.codewords[i] = candidates.index - i;
	
	codeword_list next_candidates;
	next_candidates.index = 0;
	populate_candidates(0, &candidates, &next_candidates);
	return find_best_code(&code, &next_candidates, 0);
	
}

int main(int argc, char** argv)
{
	if (argc < 3) {
		usage(argv[0]);
		exit(1);
	}

	int n, min_hd;
	if ((n 		 = atoi(argv[1])) == 0 ||
		(min_hd  = atoi(argv[2])) == 0 ||
		n > MAX_N) {
		usage(argv[0]);
		exit(1);
	}

	precompute_hd(min_hd);
	find_code(n);

	return 0;
}
